### ISSUE TITLE
Support deep population

### ISSUE THREAD

**Issue by timbur** [Automattic/mongoose#2911]
I realize this has been discussed (at least) a few times before, and it's considered a bit of an anti-pattern, but I think it's something that should be supported out of the box.  I like the simplicity of @buunguyen's [mongoose-deep-populate](https://github.com/buunguyen/mongoose-deep-populate), but the required approach (monkey-patching, etc.) makes further customization and integration with other plugins too difficult and hacky for my tastes.

For example, using the ideas behind Facebook's GraphQL (web apps should know and retrieve exactly the data it needs), a very contrived example of some query structure:
```
Comment {
  for,
  text,
  created,
  createdBy {
    name,
    avatar,
    followers.first(5) {
      name
    }
  }
}
```

And so the app would ideally be able to retrieve everything with a single request:
```
http://localhost/api/comments?conditions[for]=someId&select=for text created createdBy.name createdBy.avatar createdBy.followers.name&limit[createdBy.followers]=5
```

I think mongoose has the potential to shine here.  Ideally, the resulting query on the server might look like:
```js
Comment
  .find({for: someId})
  .select('for text created createdBy')
  .populate({
    path: 'createdBy',
    select: 'name avatar followers',
    populate: [{
      path: 'followers',
      select: 'name',
      options: { limit: 5 }
    }]
  })
  .exec();
```

All that's really needed is for mongoose to support something like this:
```js
query.populate({
  populate: [
    {
      path: ...,
      populate: [...]
    },
    {
      path: ...,
      populate: [...]
    },
    etc.
  ]
});
```

The object passed to `.populate()` should probably be constructed by something else, which would allow other plugins/middleware to alter it as necessary and would also keep things simple for mongoose.  I can't imagine it would be too difficult to implement and I'd be glad to do it myself if there are no objections.  It would certainly help me out a lot to have this in the core of mongoose rather than as some monkey-patched solution!

It's entirely possible I'm just being dumb and blind to a better way to do this.  If that's the case, please let me know.  :)

**Comment by vkarpov15** [comment]
You're welcome to give it a shot, but I imagine it would be fairly tricky. Mongoose does a lot of magic to make populate() behave in a mostly sane way, I doubt it would be a trivial task to properly support deep population.

Also, I worry that enabling this will encourage people to use mongodb for graph-like applications, which isn't really mongodb's sweet spot. In an ideal mongodb setup, you don't use `populate()` at all.

**Comment by chetverikov** [comment]
@vkarpov15 @timbur Sorry, may be I something do not understand, but I've been using the deep population... Make such requests to the API:

```php
        $params = [
            'populate' => [
                // Альбомы
                [
                    'path'   => 'albums.album',
                    'select' => 'photos description title'
                ],
                // Город
                [
                    'path'   => 'location.city',
                    'select' => 'name identity'
                ],
                //Акции
                [
                    'path'   => 'offers',
                    'select' => '-location',
                    'match'  => [
                        "schedule.end"   => [ '$gte' => time() * 1000 ],
                        "schedule.start" => [ '$lte' => time() * 1000 ]
                    ]
                ],
                // Город
                [
                    'path'   => 'menus',
                    'model'  => 'Menu'
                ],
                // Отзывы
                [
                    'path'  => 'reviews',
                    'select' => '-location',
                    'match' => [
                        'active' =>  true
                    ],
                    'options' => [
                        'populate' => 'author', // here deep populate
                        'sort' => '-created',
                        'limit' => $limit_reviews
                    ]
                ]
            ],
            'fields' => '-_-',
            'filters' => [
                '__t' => 'Food',
                'active' => true,
                'identity' => $identity,
                'location.city' => $this->_helper->city->getCity()['_id']
            ],
            'limit' => 1
        ];
```

**Comment by timbur** [comment]
What are you asking exactly?  Are you saying that works for you?  Or are you asking how to make it work?

**Comment by chetverikov** [comment]
I say this even works in version 3.8 mongoose

**Comment by chetverikov** [comment]
Without plugin

**Comment by vkarpov15** [comment]
Hmm I'd be curious to try, I've actually never attempted this. @chetverikov has done a lot of work on populate() so he would know better than me right now.

**Comment by chetverikov** [comment]
@vkarpov15 Yes, try it. I already gave an example of how I choose from your API data with deep population.

**Comment by timbur** [comment]
@chetverikov Can you go deeper?  E.g.:
```js
$params = [
    'populate' => [
        // Отзывы
        [
            'path'  => 'reviews',
            'select' => '-location',
            'match' => [
                'active' =>  true
            ],
            'options' => [
                'populate' => [           // go deeper
                    'path' => 'author',
                    'select' => 'friends',
                    'options' => [
                        'populate' => [   // and deeper
                            'path'  => 'friends',
                            'select' => 'name',
                            'options' => [
                                'populate' => 'friends',
                                'limit' => $limit_friends
                            ]
                        ]
                    ]
                ]
            ]
        ]
    ]
];
```

**Comment by chetverikov** [comment]
Yes, indeed there is such a problem. Will try to fix soon. But to be honest, to do the populate deeper than 2 times very very bad. You will have too much data in the buffer, and may exceed the call stack.

**Comment by chetverikov** [comment]
Error to be about here: https://github.com/Automattic/mongoose/blob/master/lib/model.js#L2362
In the Populate method working with options since the beginning was not correct, and now it's a real pain...

**Comment by timbur** [comment]
@chetverikov I can see how excessively deep population could be pretty resource intensive, but it would be easy enough to prevent queries from going too deep at the application level, and I still think it's a useful feature to have.  It's much easier and more efficient to make this work at the mongoose level than it would be to make it work at the application level, which was the first thing I tried.

Did you try the code I provided with the pull request?  For the sake of testing, I tried going 10+ levels deep and it seemed to work perfectly.

Side note: Personally, putting the `populate` key at the same level as `path`, `select`, etc. (rather than at `options.populate`) makes more sense to me.
